### PR TITLE
[SPARK-45452][SQL][FOLLOWUP] Simplify path check logic

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
### What changes were proposed in this pull request?

This is a follow-up of #43261 to simplify path check logic and add more test coverage.

### Why are the changes needed?

To be robust.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Pass the CIs with the newly added test case.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
Could you review this follow-up, @viirya ?

File: core/src/main/scala/org/apache/spark/util/HadoopFSUtils.scala [hunk 1]
```diff
@@ -30,6 +31,7 @@ import org.apache.spark._
 import org.apache.spark.internal.Logging
 import org.apache.spark.metrics.source.HiveCatalogMetrics
 
+
```
Comment (viirya) [inline core/src/main/scala/org/apache/spark/util/HadoopFSUtils.scala:34]:
```suggestion
```

File: core/src/main/scala/org/apache/spark/util/HadoopFSUtils.scala [hunk 2]
```diff
@@ -342,13 +345,24 @@ private[spark] object HadoopFSUtils extends Logging {
...
+      case Some(dir) if dir.equals("/_metadata") || dir.equals("/_common_metadata") => false
+      case Some(_) => true
+      case None =>
+        underscore.findFirstIn(path) match {
+          case Some(dir) if dir.equals("/_metadata/") =>
+            shouldFilterOutPath(path.replaceFirst("/_metadata", ""))
+          case Some(dir) if dir.equals("/_common_metadata/") =>
+            shouldFilterOutPath(path.replaceFirst("/_common_metadata", ""))
+          case Some(_) => true
+          case None => false
```
Comment (viirya) [inline core/src/main/scala/org/apache/spark/util/HadoopFSUtils.scala:364]: Hmm, I'm not sure if this is to simplify the logic, looks like it gets more complicated?

File: core/src/main/scala/org/apache/spark/util/HadoopFSUtils.scala [hunk 3]
```diff
@@ -342,13 +345,24 @@ private[spark] object HadoopFSUtils extends Logging {
...
-      path.endsWith("/_metadata")
-    (exclude && !include) || shouldFilterOutPathName(path)
+    if (path.contains("/.") || path.endsWith("._COPYING_")) return true
+    underscoreEnd.findFirstIn(path) match {
+      case Some(dir) if dir.equals("/_metadata") || dir.equals("/_common_metadata") => false
+      case Some(_) => true
+      case None =>
+        underscore.findFirstIn(path) match {
+          case Some(dir) if dir.equals("/_metadata/") =>
+            shouldFilterOutPath(path.replaceFirst("/_metadata", ""))
```
Comment (viirya) [inline core/src/main/scala/org/apache/spark/util/HadoopFSUtils.scala:360]: So based on updated logic, a dir looks like `"/_metadata/.ab"` is filtered out. But previously, it is not. Which one is more correct?

Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/util/HadoopFSUtils.scala:360]: New one is consistent with the existing `shouldFilterOutPathName` and correct.

File: core/src/main/scala/org/apache/spark/util/HadoopFSUtils.scala (hunk 2 above)
Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/util/HadoopFSUtils.scala:364]: Now, we don't depend on `shouldFilterOutPathName` and in the caller part. I removed the root directory to simplify this method's role.

```scala
- }.filterNot(status => shouldFilterOutPath(status.getPath.toString))
+ }.filterNot(status => shouldFilterOutPath(status.getPath.toString.substring(prefixLength)))
```

File: core/src/main/scala/org/apache/spark/util/HadoopFSUtils.scala (hunk 1 above)
Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/util/HadoopFSUtils.scala:34]: Thank you!

**Comment by dongjoon-hyun** [comment]
Thank you for review, @viirya . The PR is updated~

File: core/src/main/scala/org/apache/spark/util/HadoopFSUtils.scala [hunk 4]
```diff
@@ -342,13 +344,24 @@ private[spark] object HadoopFSUtils extends Logging {
...
-    val exclude = (path.contains("/_") && !path.contains("=")) || path.contains("/.")
-    val include = path.contains("/_common_metadata/") ||
-      path.endsWith("/_common_metadata") ||
-      path.contains("/_metadata/") ||
-      path.endsWith("/_metadata")
-    (exclude && !include) || shouldFilterOutPathName(path)
+    if (path.contains("/.") || path.endsWith("._COPYING_")) return true
+    underscoreEnd.findFirstIn(path) match {
+      case Some(dir) if dir.equals("/_metadata") || dir.equals("/_common_metadata") => false
+      case Some(_) => true
```
Comment (viirya) [inline core/src/main/scala/org/apache/spark/util/HadoopFSUtils.scala:355]: I think `/_ab=123` is not filtered out by `shouldFilterOutPathName`, but it is filtered out by this `shouldFilterOutPath`?

Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/util/HadoopFSUtils.scala:355]: Oh. You're right. Thank you. Let me fix that.

**Comment by dongjoon-hyun** [comment]
The regular expression is updated and new test cases are added. Thank you so much, @viirya .
```
[info] HadoopFSUtilsSuite:
[info] - HadoopFSUtils - file filtering (14 milliseconds)
[info] - SPARK-45452: HadoopFSUtils - path filtering (1 millisecond)
[info] Run completed in 906 milliseconds.
[info] Total number of tests run: 2
[info] Suites: completed 1, aborted 0
[info] Tests: succeeded 2, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.
[success] Total time: 15 s, completed Oct 9, 2023, 12:39:34 AM
```

**Review by viirya — APPROVED** [review]

**Comment by viirya** [comment]
Looks good to me now.

File: core/src/test/scala/org/apache/spark/util/HadoopFSUtilsSuite.scala [hunk 5]
```diff
@@ -30,4 +30,36 @@ class HadoopFSUtilsSuite extends SparkFunSuite {
...
+    assert(HadoopFSUtils.shouldFilterOutPath("/.ab/cde"))
+    assert(HadoopFSUtils.shouldFilterOutPath("/.ab/_metadata/1"))
+    assert(HadoopFSUtils.shouldFilterOutPath("/ab/.cde"))
+    assert(HadoopFSUtils.shouldFilterOutPath("/ab/.cde/fg"))
+    assert(HadoopFSUtils.shouldFilterOutPath("/ab/.cde/_metadata"))
+    assert(HadoopFSUtils.shouldFilterOutPath("/ab/.cde/_common_metadata"))
+    assert(HadoopFSUtils.shouldFilterOutPath("/ab/.cde/year=2023/month=10/day=8/hour=13"))
+    assert(HadoopFSUtils.shouldFilterOutPath("/x/.hidden/part=__HIVE_DEFAULT_PARTITION__"))
+    assert(HadoopFSUtils.shouldFilterOutPath("/a._COPYING_"))
+    // Case 3: Understored paths (except metadata paths of Case 1)
```
Comment (viirya) [inline core/src/test/scala/org/apache/spark/util/HadoopFSUtilsSuite.scala:57]: Understored? Underscored?

**Comment by dongjoon-hyun** [comment]
Merged to master for Apache Spark 4.0.0. Thank you, @viirya .

**Pull request closed without merging**